plugin/forward/dnstap.go

```diff
@@ -1,61 +1,60 @@
 package forward
 
 import (
-	"context"
+	"net"
+	"strconv"
 	"time"
 
-	"github.com/coredns/coredns/plugin/dnstap"
 	"github.com/coredns/coredns/plugin/dnstap/msg"
 	"github.com/coredns/coredns/request"
 
 	tap "github.com/dnstap/golang-dnstap"
 	"github.com/miekg/dns"
 )
 
-func toDnstap(ctx context.Context, host string, f *Forward, state request.Request, reply *dns.Msg, start time.Time) error {
-	tapper := dnstap.TapperFromContext(ctx)
-	if tapper == nil {
-		return nil
-	}
+// toDnstap will send the forward and received message to the dnstap plugin.
+func toDnstap(f *Forward, host string, state request.Request, reply *dns.Msg, start time.Time) {
 	// Query
-	b := msg.New().Time(start).HostPort(host)
+	tm := new(tap.Message)
+	msg.SetQueryTime(tm, start)
+	ip, p, _ := net.SplitHostPort(host)     // this is preparsed and can't err here
+	port, _ := strconv.ParseUint(p, 10, 32) // same here
+
 	opts := f.opts
-	t := ""
+	t := state.Proto()
 	switch {
 	case opts.forceTCP: // TCP flag has precedence over UDP flag
 		t = "tcp"
 	case opts.preferUDP:
 		t = "udp"
```

Comment:
this is not quite reliable. Forward may try udp and finally use TCP. The `opts` from the line https://github.com/coredns/coredns/blob/02f52b5d271d47a035c5432cca4fcf4174c023fa/plugin/forward/forward.go#L126 has more accurate value, since it is updated after UDP failure

Reply:
and done.